relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	uuid "github.com/satori/go.uuid"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+
+	"github.com/pingcap/dm/pkg/binlog/event"
+	"github.com/pingcap/dm/pkg/binlog/reader"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/relay/common"
+)
+
+// checkBinlogHeaderExist checks if the file has a binlog file header.
+func checkBinlogHeaderExist(filename string) (bool, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	buff := make([]byte, fileHeaderLen)
+	n, err := f.Read(buff)
+	if n == 0 && err == io.EOF {
+		return false, nil // empty file
+	} else if n != fileHeaderLen {
+		return false, errors.Errorf("binlog file %s has no enough data, only got % X", filename, buff[:n])
+	} else if err != nil {
+		return false, errors.Annotate(err, "read binlog header")
+	}
+
+	if !bytes.Equal(buff, replication.BinLogFileHeader) {
+		return false, errors.Errorf("binlog file %s header not valid, got % X, expect % X", filename, buff, replication.BinLogFileHeader)
+	}
+	return true, nil
+}
+
+// checkFormatDescriptionEventExist checks if the file has a valid FormatDescriptionEvent.
+func checkFormatDescriptionEventExist(filename string) (bool, error) {
+	// FormatDescriptionEvent always follows the binlog file header
+	exist, err := checkBinlogHeaderExist(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		return false, errors.Errorf("no binlog file header at the beginning for %s", filename)
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// check whether only the file header
```

Comment:
I think we can save this logic

Reply:
we need to diff two kinds of EOF when parsing the FormatDescriptionEvent:
1. only the file header
2. part of the FormatDescriptionEvent header

if we save this logic, we need another logic to diff the returned values of `ParseSingleEvent`.

Reply:
ok